scripts/copydocs.py

```diff
@@ -124,41 +147,93 @@ def get_references(doc, result=None):
     return result
 
+
+class KeyVersionPair(namedtuple('KeyVersionPair', 'key version')):
+    """Helper class to store uri's like /works/OL1W?v=2"""
+
+    @staticmethod
+    def from_uri(uri):
+        """
+        :param str uri: either something like /works/OL1W, /books/OL1M?v=3, etc.
+        :rtype: KeyVersionPair
+        """
+
+        if '?v=' in uri:
+            key, version = uri.split('?v=')
+        else:
+            key, version = uri, None
+        return KeyVersionPair._make([key, version])
+
+    def to_uri(self):
+        """
+        :rtype: str
+        """
+        uri = self.key
+        if self.version:
+            uri += '?v=' + self.version
+        return uri
+
+    def __str__(self):
+        return self.to_uri()
+
+
 def copy(src, dest, keys, comment, recursive=False, saved=None, cache=None):
+    """
+    :param Disk or OpenLibrary src: where we'll be copying form
+    :param Disk or OpenLibrary dest: where we'll be saving to
+    :param typing.List[str] keys:
+    :param str comment: commit to writing when saving the documents
+    :param bool recursive:
+    :param typing.Set[str] or None saved: keys saved so far
+    :param dict or None cache:
+    """
     if saved is None:
         saved = set()
     if cache is None:
         cache = {}
 
     def get_many(keys):
         docs = marshal(src.get_many(keys).values())
-        # work records may contain excepts, which reference the author of the excerpt.
+        # work records may contain excerpts, which reference the author of the excerpt.
         # Deleting them to prevent loading the users.
-        # Deleting the covers and photos also because they don't show up in the dev instance.
         for doc in docs:
             doc.pop('excerpts', None)
-            #doc.pop('covers', None)
-            #doc.pop('photos', None)
 
             # Authors are now with works. We don't need authors at editions.
             if doc['type']['key'] == '/type/edition':
                 doc.pop('authors', None)
 
         return docs
 
-    def fetch(keys):
+    def fetch(uris):
+        """
+        :param typing.List[str] uris:
+        :rtype: typing.List[dict or web.storage]
+        """
         docs = []
 
-        for k in keys:
-            if k in cache:
-                docs.append(cache[k])
+        key_pairs = list(map(KeyVersionPair.from_uri, uris))
+
+        for pair in key_pairs:
+            if pair.key in cache:
+                docs.append(cache[pair.key])
 
-        keys = [k for k in keys if k not in cache]
-        if keys:
-            print("fetching", keys)
-            docs2 = get_many(keys)
+        key_pairs = [pair for pair in key_pairs if pair.to_uri() not in cache]
+
+        unversioned_keys = [pair.key for pair in key_pairs if pair.version is None]
+        versioned_to_get = [pair for pair in key_pairs if pair.version is not None]
+        if unversioned_keys:
+            print("fetching", unversioned_keys)
```

Comment:
we may want to rm this print (and use logs if necessary)

Reply:
If you don't mind, I'd rather punt this to another commit. The output is not super noisy, and I'd rather not introduce logging as part of this PR cause that would be a pretty hefty change.